_posts/2017-10-30-user-secrets.md

```diff
@@ -0,0 +1,89 @@
+<br/>
+In my case my user secret looks like:
+```
+// ~/.microsoft/usersecrets/serviceConfigSecretId/secrets.json
+{
+  "SecretServiceConfig": {
+    "address": "https://somewhere-secret.com",
+    "key": "super-secret-key"
+  }
+}
+```
+
+Then, the user secret must be referenced in the project file:
+```xml
+<PropertyGroup>
+    <TargetFramework>netcoreapp2.0</TargetFramework>
+    <UserSecretsId>serviceConfigSecretId</UserSecretsId>
+</PropertyGroup>
+```
+
+With that in place, the user secret can now be used in code.
+
+
+##### Add the User Secret to the Configuration at run time
+
+Ultimately, when my application will be deployed, the sensitive data will be injected via the configuration at start time. So what I want is to be able to add the User Secrets to the configuration at start time but only in development mode. The way (or at least one way) to do it to add the user secret to the C# Configuration object when it is built:
+```
+public class Startup
+{
+    public Startup(IHostingEnvironment env)
+    {
+        var builder = new ConfigurationBuilder()
+            .SetBasePath(env.ContentRootPath)
+            .AddJsonFile("appsettings.json");
+
+        if (env.IsDevelopment())
+        {
+            builder.AddUserSecrets<Startup>();
+        }
+
+        Configuration = builder.Build();
+    }
+```
+An interesting thing to note here is that one can mix configuration sources. In my case, my application configuration comes from two sources: one is the appsettings.json file stored with my code; the other is the data stored in the user secrets. This is super useful!
+
+Once the configuration object is built, it doesn't matter where the data came from, you can access it as usual by calling `Configuration.GetSection("SecretServiceConfig")`. This will return a ConfigurationSection object with all the data.
+
+
+##### Using Option to access Configuration
+
+The Configuration object can be used directly but a more elegant way to access configuration is using what Microsoft calls the option pattern. It basically helps match configuration section to C# classes. I won't describe how I used it in this post as it is out of scope but it is worth knowing about it, you can check out this [article](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration?tabs=basicconfiguration) about Configuration, they mention it.
+
+<br/>
+
+So Secret Manager is very useful when developping ASP.NET Core applications locally. There are other ways of achieving the same things (like using environment variables) but I like this solution as it is very straighforward to configure. There is only one thing that bothers me: the extra `if` statement in code to execute extra operations in development. I don't like it very much, I prefer when the code path remains the same no matter what environments the code is running in. However, that will do for now!
```

Comment:
Do you need that `if` statement? Perhaps I've misunderstood under what circumstances wouldn't you want to inject the secret?

Reply:
You asking that is not a good sign, it's the whole point of my article.
Secret Manager is there to help you when developing locally. Once in production you inject the secrets differently (config file or environment variables)

Reply:
Ah, ok! It's the last bit I didn't get - that you would use a different method of injecting them in production

Reply:
Yeah, I am adding more details on that, push coming...